src/script/sigcache.h

```diff
@@ -16,13 +17,40 @@ static const unsigned int DEFAULT_MAX_SIG_CACHE_SIZE = 40;
 
 class CPubKey;
 
+/**
+ * A thread safe map which caches mid state witness signature hash calculation by transaction id
+ */
+class CachedHashesMap
+{
+private:
+    std::map<uint256, CachedHashes> map;
+    CCriticalSection cs;
+public:
+    bool TryGet(uint256 txId, CachedHashes* hashes)
+    {
+        LOCK(cs);
+        if(!map.count(txId))
+            return false;
+        *hashes = map[txId];
+        return true;
+    }
+    bool TrySet(uint256 txId, const CachedHashes& hashes)
+    {
+        LOCK(cs);
+        if(map.count(txId))
```

Comment:
You could avoid two look ups by comparing the size of the container before and after instead.

Aka:
``` cpp
auto sizeBefore = map.size();
auto pair = map.insert(txId, hashes);
return map.size() != sizeBefore;
```